### ISSUE TITLE
[REFACTOR] Opus5 평가 QA 결과

### ISSUE THREAD

**Issue by rokpolar** [memorybridge-team/vos-memory-translator-nonlinear#7]
### 해당 프로젝트 내용을 Opus5 High에서 평가한 결과 및 발견한 수정사항 입니다.

의미가 아니라 코드 자체의 결함으로 확인된 것이 6건이고, 그중 3건은 지금 실험을 돌리면 논문 수치를 직접 왜곡합니다. 아래 수치는 모두 실제로 실행해서 얻은 값입니다.

결과에 영향을 주는 결함
1. 학습형 translator가 bfloat16 → float32로 조용히 승격되어 handoff 전달량이 정확히 2배로 보고됩니다.

_feature_head가 입력을 Linear weight dtype(fp32)으로 캐스팅하고, 그 fp32 텐서가 그대로 with_continuous에 저장됩니다.

translators.py
Ln 379–385
def _feature_head(self, tensor: torch.Tensor) -> torch.Tensor:
    calibrated = tensor.to(
        device=self.feature[0].weight.device,
        dtype=self.feature[0].weight.dtype,
    )
handoff_bytes()는 numel * element_size()로 세므로, 같은 11-record payload가 Direct Copy는 5,778,641 bytes(bf16), Linear/Residual-MLP는 11,545,809 bytes(fp32)로 나옵니다. 앞의 값은 PROJECT_CONTEXT.md §21에 기록된 실측치와 정확히 일치하므로, 이 2배 차이는 방법의 특성이 아니라 dtype 사고입니다. 게다가 inject_sam2_canonical_state는 device만 옮기고 dtype은 건드리지 않아서, Target이 native에서는 bf16인 maskmem_features를 nonlinear 경로에서만 fp32로 받게 됩니다. Direct 군과 Nonlinear 군 사이에 통제되지 않은 차이가 생깁니다.

2. fit_affine_closed_form이 float64로 푼 해를 입력 dtype으로 되돌려 버려서 Ridge/Moment-Matched baseline이 실제보다 약해집니다.

translators.py
Ln 129–130
output_dtype = inputs.dtype if inputs.is_floating_point() else torch.float32
return AffineMap(weight.to(output_dtype), bias.to(output_dtype))
spatial memory가 bf16이므로 64×64 weight 행렬이 가수 8비트로 잘리고, AffineMap.__call__도 bf16으로 matmul합니다. y = 2x + 1이라는 정확한 affine 관계를 학습시켜 봤더니 fp32는 최대 오차 1e-6, bf16은 0.071이었고, SAM 2 형태(64ch/64×64)의 bf16 pair에서는 0.0156이었습니다. nonlinear가 필요한지를 판정하는 기준선이 부당하게 낮아지는 방향의 편향입니다.

3. --component-policy spatial_pointer가 full과 완전히 같은 결과를 냅니다.

v1.1에서 presence를 번역 대상에서 뺀 뒤 _ALLOWED_COMPONENTS에는 spatial_memory와 object_pointer만 남았는데, spatial_pointer는 그 둘을 모두 선택합니다. 출력 텐서가 bit-identical임을 확인했습니다. 그런데 evaluate_nonlinear_collection.py는 둘에 다른 evaluation_id와 다른 label을 붙이므로, 재개 캐시가 이를 별개 방법으로 보고 같은 결과를 두 번 저장합니다.

같은 이유로 RidgeDirectPresenceTranslator와 --presence-policy 플래그도 완전한 no-op입니다. RidgeStateTranslator.translate가 이미 presence_logits=source.presence_logits.clone()을 하고 있어서, ridge와 direct 두 선택지가 진단용 presence까지 포함해 모든 필드에서 동일한 값을 냅니다.

그 외 실제 버그
4. benchmark_translation(warmup=0)은 UnboundLocalError로 죽습니다. result가 warmup 루프 안에서만 바인딩되는데 루프 밖에서 result.spatial_memory.is_cuda를 읽습니다. 현재 호출부가 전부 warmup=2라 드러나지 않은 잠복 버그입니다.

5. 갤러리 HTML의 "DAVIS GT" 패널이 실제로는 Target-native를 보여줍니다. build_baseline_suite_gallery.py는 GT 오버레이를 images["gt"]로 저장해 놓고, JS show()에서 #gt의 src를 r.images.target_native로 지정합니다. #gt와 #native 두 칸이 같은 이미지입니다. mp4 경로는 정상입니다.

6. MemoryAttentionProbe.detach()가 monkey patch를 제거하지 않습니다. 원래 bound method를 인스턴스 속성으로 다시 대입만 해서, class method를 가리는 인스턴스 속성이 영구히 남고 predictor ↔ bound method 참조 순환이 생깁니다. del self.predictor._prepare_memory_conditioned_features가 맞습니다.

사소한 것
sam2_state.py의 _move_nested_tensors는 자기 재귀 외에 참조가 없는 죽은 코드입니다. _regular_frames의 int(round(q * (len-1)))은 Python의 banker's rounding 때문에 round(0.5) == 0이라 eligible이 3개일 때 q=0.25가 q=0과 같은 프레임을 고릅니다(결정적이긴 합니다). run_cached_baseline_suite.py의 end_frame = max(frames) - 1은 DAVIS 마지막 프레임 제외 규약이 맞지만 주석이 없어 off-by-one처럼 읽힙니다. ruff는 이 외에 44건을 냈는데 전부 UP035/UP037 같은 스타일이었고, B023 1건은 루프 안에서 즉시 호출되므로 실제 문제가 아닙니다.

sam2_video_predictor.py가 저장 시점에 maskmem_features를 무조건 bfloat16으로 캐스팅하고 있습니다. 이게 두 해법의 판단을 갈라놓는 핵심이라, 실제 결과를 측정하겠습니다.

조사 결과, 두 방식은 대등한 선택지가 아니라 A가 B를 포함하는 관계입니다. B 단독으로는 지표만 고쳐지고 실험 타당성 문제가 그대로 남습니다. 코드는 수정하지 않았고 작업 트리는 깨끗합니다.

먼저: upstream이 강제하는 계약
로컬 ~/sam2가 pinned commit 2b90b9f와 일치해서 원본을 직접 확인했습니다. 저장 경로 두 곳 모두 무조건 캐스팅합니다.

sam2_video_predictor.py
Ln 780–782
if maskmem_features is not None:
    maskmem_features = maskmem_features.to(torch.bfloat16)
    maskmem_features = maskmem_features.to(storage_device, non_blocking=True)
반면 읽기 경로는 device만 옮기고 dtype은 건드리지 않은 채 torch.cat으로 합칩니다.

sam2_base.py
Ln 573–576
# "maskmem_features" might have been offloaded to CPU in demo use cases,
# so we load it back to GPU (it's a no-op if it's already on GPU).
feats = prev["maskmem_features"].to(device, non_blocking=True)
to_cat_memory.append(feats.flatten(2).permute(2, 0, 1))
그리고 이 프로젝트는 autocast를 쓰지 않습니다(rg autocast 결과는 PE factory의 is_autocast_enabled 프로브 한 곳뿐).

그래서 nonlinear를 주입하면 실제로 벌어지는 일
switch+1 프레임에서는 주입된 fp32 record만 모이므로 memory가 전부 fp32입니다. switch+2부터는 주입된 fp32 history와 Target이 새로 인코딩한 bf16 record가 섞이는데, torch.cat은 에러 없이 조용히 fp32로 승격합니다(측정 확인). autocast가 없으므로 memory attention은 두 경우 모두 fp32로 실행되고, 결국 Direct 군은 bf16 memory를, Nonlinear 군은 fp32 memory를 같은 attention에 넣게 됩니다. downstream J&F 차이에 "번역 품질"과 "메모리 수치 정밀도"가 섞여 들어갑니다.

여기서 중요한 점은, 지금까지 통과한 runtime gate가 이 경로를 한 번도 실행하지 않았다는 것입니다. self-injection과 Direct Copy는 canonical state의 native bf16을 그대로 보존하므로 bit-exact가 나왔습니다. 테스트 53개도 전부 fp32 synthetic state라 dtype을 검증하는 assertion이 하나도 없습니다. 이 결함은 RunPod에서 nonlinear를 처음 돌리는 순간 처음 나타납니다.

방식 A — translator 출력을 native bf16으로 되돌리기
제가 처음 우려했던 "forward 안에서 캐스팅하면 학습이 깨진다"는 틀렸습니다. .to(bfloat16)은 미분 가능해서 gradient가 그대로 흐릅니다. 실측 grad norm 변화 0.22%로, 학습 경로에 제약이 없습니다.

정확도 비용도 작습니다. 예측을 bf16으로 양자화했을 때 MSE가 2.7% 증가하는데, 타깃 자체가 이미 bf16이라 레퍼런스의 가수가 8비트뿐입니다. 즉 이건 정보 손실이 아니라 예측을 타깃과 같은 격자에 올리는 비용입니다.

나머지 영향은 전부 이득 쪽입니다. handoff bytes가 Direct와 같은 기준(5,778,641)이 되어 방법 간 비교가 성립하고, PROJECT_CONTEXT.md §21에 기록된 값을 재해석할 필요가 없습니다. test_handoff_bytes_counts_only_injected_payload는 element_size()로 재계산하는 방식이라 그대로 통과합니다. 부수적으로 paired/case cache 크기가 절반이 되는데, shard 하나가 166MB급인 걸 감안하면 무시할 수준이 아닙니다.

방식 B — fp32 유지 + wire dtype 기준으로 바이트 계산
핵심 한계는 주입 비대칭이 그대로 남는다는 것입니다. 지표만 고치고 실제 텐서는 fp32로 주입되므로 위의 mixed-dtype cat과 fp32 attention 경로가 유지됩니다. 6건 중 "보고 수치" 절반만 해결됩니다.

게다가 손대야 할 범위가 A보다 넓습니다. handoff_bytes()와 contract_dict()에 wire dtype 개념을 추가해야 하고, tests/test_state_transfer.py:258을 고쳐야 하며, 이미 기록된 수치들을 재해석해야 합니다 — PROJECT_CONTEXT.md §21, reports/runtime/2026-09-21_task06_edge_case_and_direct_injection/small_to_base_direct.json, reports/runtime/2026-09-20_small_base_runtime_inventory/pair_report.json과 해당 README. 그리고 I/O 계약에 dtype 정책 조항이 새로 생기므로 계약 문서 128줄의 규칙에 따라 버전을 올려야 합니다.

판단
A를 권합니다. B가 해결하는 것을 A가 모두 해결하면서 실험 타당성 문제까지 없애고, 기존 기록과 테스트를 건드리지 않으며, upstream의 저장 dtype 계약과도 일치합니다.

덧붙이면 이건 사실 새로운 설계 결정이 아닙니다. 동결된 v1.1 계약 문서가 이미 이 단계를 규정하고 있습니다.

small_base_state_io_contract.md
Ln 52–52
Export 직전에는 Source가 `non_blocking=True`로 GPU→CPU offload한 최신 memory가 완전히 도착하도록 producer CUDA device를 한 번 동기화한다. 그 다음 Injection 직전에는 Target runtime 정책에 따라 translated `spatial_memory`를 Target `storage_device`로, translated `object_pointer`와 Target-generated PE를 Target compute device로 이동한다. Source `pred_masks`와 `presence_logits`는 Target runtime에 이동·주입하지 않는다. Source device 문자열은 provenance일 뿐 Target 장치 설정을 덮어쓰지 않는다. Runtime에서 확인한 dtype은 spatial `bfloat16`, pointer/presence `float32`이며 translator 내부 정밀도와 최종 cast는 학습 config에 기록한다.
"translator 내부 정밀도와 최종 cast는 학습 config에 기록한다"는 조항이 있는데, 코드에 그 최종 cast가 구현되어 있지 않은 상태입니다. 즉 계약 변경이 아니라 미구현 조항의 이행입니다.
